File: course/src/12_equality/hashcode/Person.java
public class Person {
    private String firstName;
    private String lastName;

    public Person(String firstName, String lastName) {
        this.firstName = firstName;
        this.lastName = lastName;
    }

    @Override
    public boolean equals(Object obj) {
        if (!(obj instanceof Person)) return false;
        Person that = (Person) obj;
        return this.lastName.equalsIgnoreCase(that.lastName);
    }

    @Override
    public int hashCode() {
        return lastName.toUpperCase().hashCode();
    }

    public static void main(String[] args) {
        Person person1 = new Person("Alice", "Smith");
        Person person2 = new Person("Bob", "smith");
        Person person3 = new Person("Charlie", "Johnson");

        System.out.println("person1 equals person2: " + person1.equals(person2));
        System.out.println("person1 hashCode: " + person1.hashCode());
        System.out.println("person2 hashCode: " + person2.hashCode());
        
        System.out.println("\nperson1 equals person3: " + person1.equals(person3));
        System.out.println("person1 hashCode: " + person1.hashCode());
        System.out.println("person3 hashCode: " + person3.hashCode());
    }
}

File: course/src/12_equality/hashcode/person.py
class Person:
    def __init__(self, first_name: str, last_name: str):
        self.first_name = first_name
        self.last_name = last_name

    def __eq__(self, other):
        if not isinstance(other, Person):
            return False
        return self.last_name.lower() == other.last_name.lower()

    def __hash__(self):
        return hash(self.last_name.lower())

    def __repr__(self):
        return f"Person(first_name='{self.first_name}', last_name='{self.last_name}')"


# Testing the implementation
person1 = Person("Alice", "Smith")
person2 = Person("Bob", "smith")
person3 = Person("Charlie", "Johnson")

# Print equality and hash results
print(f"person1 == person2: {person1 == person2}")
print(f"hash(person1): {hash(person1)}")
print(f"hash(person2): {hash(person2)}")

print("\nperson1 == person3: {person1 == person3}")
print(f"hash(person1): {hash(person1)}")
print(f"hash(person3): {hash(person3)}")

